base/common/src/main/java/org/dogtagpki/acme/ACMEIdentifier.java

```diff
@@ -44,6 +44,92 @@ public void setValue(String value) {
         this.value = value;
     }
 
+    /**
+     * Validate syntax of identifier.
+     *
+     * @throws IllegalArgumentException if either type or value is null.
+     * @return a result that is either OK, or a failure with an
+     *         appropriate ACMEError attached.
+     */
+    public ValidationResult validateSyntax() {
+        if (type == null) {
+            throw new IllegalArgumentException("Programming error: type is null");
+        } else if (value == null) {
+            throw new IllegalArgumentException("Programming error: value is null");
+        }
+        switch (type) {
+            case "dns":
+                return validateSyntaxDNS();
+            default:
+                ACMEError error = new ACMEError();
+                error.setType("urn:ietf:params:acme:error:unsupportedIdentifier");
+                error.setDetail("Unsupported identifier type: " + type);
+                return ValidationResult.fail(error);
+        }
+    }
+
+    /**
+     * Validate DNS identifier.
+     *
+     * Validates per Section 3.5 of RFC 1034 and Section 2.1 of
+     * RFC 1123, and the additional rules of RFC 5280 Section 4.2.1.6.
+     *
+     * It is a precondition that type == "dns" and value != null.
+     * Wildcard as first label is valid syntax.  Wildcard
+     * _policy_ is checked elsewhere.
+     */
+    private ValidationResult validateSyntaxDNS() {
```

Comment:
Although identifier validation is well-defined in RFCs, and RFCs generally don't change, the implementation might occasionally change. For example, we might decide later to provide more specific error messages, or we might want to provide localized error messages, or we might support other types of identifiers.  So rather than making the POJO more complicated I'd prefer we move this into the `ACMEPolicy.validateIdentifier()` and merged with the existing wildcard validation:
https://github.com/dogtagpki/pki/blob/master/base/acme/src/main/java/org/dogtagpki/acme/server/ACMEPolicy.java#L28

Reply:
Also, our interpretation of the RFCs might change, or might be different than that of the admin actually deploying the service. Or there might be other RFCs or other things that we have not considered. See also:
https://stackoverflow.com/questions/8724954/what-is-the-maximum-number-of-characters-for-a-host-name-in-unix
So I really think this code doesn't belong in the ACME POJO class.

Reply:
I agree to move the syntax validation behaviour out of the `ACMEIdentifier` class.  But it will go somewhere else; not to `ACMEPolicy.validateIdentifier`.  Syntax validation and policy checks are separate concerns.  In fact, keeping their implementations separate should make it *easier* to do the kinds of changes you discussed above.